synapse/module_api/__init__.py

```diff
@@ -1452,6 +1452,65 @@ async def get_monthly_active_users_by_service(
             start_timestamp, end_timestamp
         )
 
+    async def create_room(
+        self,
+        user_id: str,
+        config: JsonDict,
+        ratelimit: bool = True,
+        creator_join_profile: Optional[JsonDict] = None,
+    ) -> Tuple[dict, int]:
+        """Creates a new room.
+
+        Args:
+            user_id:
+                The user who requested the room creation.
+            config : A dict of configuration options. See "Request body" of:
+                https://spec.matrix.org/latest/client-server-api/#post_matrixclientv3createroom
+            ratelimit: set to False to disable the rate limiter.
+
+            creator_join_profile:
+                Set to override the displayname and avatar for the creating
+                user in this room. If unset, displayname and avatar will be
+                derived from the user's profile. If set, should contain the
+                values to go in the body of the 'join' event (typically
+                `avatar_url` and/or `displayname`.
+
+        Returns:
+                First, a dict containing the keys `room_id` and, if an alias
+                was, requested, `room_alias`. Secondly, the stream_id of the
+                last persisted event.
+        Raises:
+            SynapseError if the user does not exist, room ID couldn't be stored, or
+            something went horribly wrong.
+            ResourceLimitError if server is blocked to some resource being
+            exceeded.
+        """
+        user_info = await self.get_userinfo_by_id(user_id)
+        if user_info is None:
+            raise SynapseError(400, f"User ({user_id}) not found")
+
+        if user_info.appservice_id is not None:
+            app_service = self._store.get_app_service_by_id(
+                str(user_info.appservice_id)
+            )
+        else:
+            app_service = None
+
+        requester = create_requester(
```

Comment:
Based on what you said [here](https://github.com/matrix-org/synapse/pull/11147#issuecomment-1200851379) (regarding shadow bans and `create_requester`), do you think the way in which `create_requester` is being invoked here (and the code above it) is a bit redundant? I.e. Do you think we can get away with something like [this](https://github.com/matrix-org/synapse/blob/e17e5c97e0d2be7a82c782e8ef9774e682968c7b/synapse/module_api/__init__.py#L966), do away with the code above, and add a pre-condition docstring comment that the user must exist?

Reply:
I think it should just check if the user ID is valid and local. There are valid use cases for puppetting a user that doesn't actually exist (Synapse actually does that for server notices) so I'm not sure the requirement on the user existing is necessary.

Reply:
I noticed (during development of my use-case) that Synapse allowed the module to puppet a user that didn't exist. At the time I thought it was due to me misusing Synapse internals (i.e. not doing the necessary validation) and there being a lack of constraints in the database layer (to enforce the validation), but it seems you're saying that is by design. That is an unexpected lesson for me, but given that it's known to Synapse core devs, I've altered the module API method as per your suggestion: 916c583

Reply:
> but it seems you're saying that is by design

To be clear, I'm not 100% sure allowing modules to do it was intentional, but I also don't really see a reason to prevent this.